runtime/src/account_rent_state.rs

```diff
@@ -0,0 +1,159 @@
+use {
+    enum_iterator::IntoEnumIterator,
+    log::*,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        native_loader,
+        rent::Rent,
+        sysvar,
+        transaction::{Result, TransactionError},
+        transaction_context::TransactionContext,
+    },
+};
+
+#[derive(Debug, PartialEq, IntoEnumIterator)]
+pub(crate) enum RentState {
+    Uninitialized, // account.lamports == 0
+    NativeOrSysvar,
+    RentPaying, // 0 < account.lamports < rent-exempt-minimum
+    RentExempt, // account.lamports >= rent-exempt-minimum
+}
+
+impl RentState {
+    pub(crate) fn from_account(account: &AccountSharedData, rent: &Rent) -> Self {
+        if account.lamports() == 0 {
+            Self::Uninitialized
+        } else if native_loader::check_id(account.owner()) || sysvar::is_sysvar_id(account.owner())
```

Comment:
I think it's sufficient to just have a special case for executable accounts here. Sysvars are subject to rent rules as of this feature:
```
BKCPBQQBZqggVnFso5nQ8rQ4RwwogYwjuUt9biBjxwNF | active since slot 104976000 | collect rent from accounts owned by sysvars
```

Reply:
Actually, I think that as of my last commit (only checking writable accounts), this special case isn't needed at all. Native programs should always be demoted to read-only, and txs with writable executables should only make it here if the upgradeable loader is present.
Does that make sense?

Reply:
Yup, that makes sense! Could you maybe add an assertion in there for that case to prevent that assumption from breaking this?

Reply:
For sure. To be clear, you mean assert in `check_rent_state` that an account with `Some` info is not a sysvar or owned by the native loader, right?